### ISSUE TITLE
Avoid pip install in every notebook in one project?

### ISSUE THREAD

**Issue by kolibril13** [jupyterlite/pyodide-kernel#48]
In order to run a notebook from the ipyreact JupyterLite demo at https://widgetti.github.io/ipyreact,
it requires the line `%pip install -q ipyreact`.
This is currently needed in every notebook, and I think it would be great if this line would only be needed in one notebook, or even not at all.   
Therefore, my question: Is there yet a way to avoid having this line in every notebook using the pyodide kernel?

**Comment by kolibril13** [comment]
I've tried to use the xeus kernel for ipyreact, but that throws an CI error: 
https://github.com/kolibril13/ipyreact-testing-xeus/actions/runs/5145098489/jobs/9262258540
```
error    libmamba Could not solve for environment specs
    The following package could not be installed
    └─ ipyreact   does not exist (perhaps a typo or a missing channel).
```

**Comment by martinRenou** [comment]
It looks like ipyreact is not available on conda-forge, could be good to have it though ! @maartenbreddels friendly ping :D

**Comment by jtpio** [comment]
Ah looks like it's not packaged for conda. We should open a PR on https://github.com/conda-forge/staged-recipes to add it.

**Comment by maartenbreddels** [comment]
I suffer from conda-forge packaging fatigue :)

**Comment by maartenbreddels** [comment]
What is the reason conda-forge packages work, but pip packages don't?

**Comment by martinRenou** [comment]
> I suffer from conda-forge packaging fatigue :)

😅 if it's not a pure Python package, can you please make a PR to emscripten-forge as well? (kidding).

https://github.com/conda-forge/staged-recipes/pull/22993 🚀

**Comment by jtpio** [comment]
> I suffer from conda-forge packaging fatigue :)

Well often you can generate most of the recipe with `grayskull`: https://github.com/conda/grayskull

**Comment by maartenbreddels** [comment]
Yes, but it still involves manual steps, especially when requirements/pinning changes. I've streamlined so much of the release process in all my projects, but also supporting conda-forge just sucks the energy out of me for letting me do things manually that feel unnecessary. I like conda/mamba and conda-forge a lot, but the duplicate work that is not automated (but could) just feels very unproductive.

**Comment by juntyr** [comment]
See https://github.com/jupyterlite/pyodide-kernel/pull/167#issuecomment-2655900778 for how to preload packages from the pyodide lockfile where the package list comes from a URL query parameter. With https://github.com/jupyterlite/pyodide-kernel/pull/172 it would also be possible to install packages with micropip before the kernel finishes initialization.

**Comment by martinRenou** [comment]
Using jupyterlite-xeus with xeus-python is the simplest way to achieve this:

https://jupyterlite-xeus.readthedocs.io/en/latest/environment.html
